File: problems/019-reproducible-build-order/solutions/javascript/main.js
import * as std from "std";
const a = std.in.readAsString().trim().split(/\s+/);
let p = 0,
  n = Number(a[p++]),
  m = Number(a[p++]);
/** @type {string[]} */
const names = Array.from({ length: n }, () => a[p++]);
/** @type {number[]} */
const nameOrder = Array.from({ length: n }, (_, i) => i).sort((x, y) =>
  names[x] < names[y] ? -1 : names[x] > names[y] ? 1 : 0
);
/** @type {number[][]} */
const g = Array.from({ length: n }, () => []);
const deg = new Int32Array(n);
let bad = 0;
/** @param {string} name @returns {number} */
const findId = (name) => {
  let lo = 0, hi = n;
  while (lo < hi) {
    const mid = (lo + hi) >> 1;
    if (names[nameOrder[mid]] < name) lo = mid + 1;
    else hi = mid;
  }
  return lo < n && names[nameOrder[lo]] === name ? nameOrder[lo] : -1;
};
for (let i = 1; i <= m; i++) {
  const x = findId(a[p++]), y = findId(a[p++]);
  if (x < 0 || y < 0) { if (!bad) bad = i; }
  else {
    g[y].push(x);
    deg[x]++;
  }
}
if (bad) std.out.puts(`INVALID DANGLING ${bad}\n`);
else {
  /** @type {number[]} */
  const h = [];
  /** @param {number} x @param {number} y @returns {boolean} */
  const less = (x, y) => names[x] < names[y];
  /** @param {number} x @returns {void} */
  const push = (x) => {
    let i = h.length;
    h.push(x);
    while (i) {
      let q = (i - 1) >> 1;
      if (!less(h[i], h[q])) break;
      [h[i], h[q]] = [h[q], h[i]];
      i = q;
    }
  };
  /** @returns {number} */
  const pop = () => {
    const r = h[0], x = /** @type {number} */ (h.pop());
    if (h.length) {
      h[0] = x;
      for (let i = 0;;) {
        let l = i * 2 + 1, b = i;
        if (l < h.length && less(h[l], h[b])) b = l;
        if (l + 1 < h.length && less(h[l + 1], h[b])) b = l + 1;
        if (b === i) break;
        [h[i], h[b]] = [h[b], h[i]];
        i = b;
      }
    }
    return r;
  };
  for (let i = 0; i < n; i++) if (!deg[i]) push(i);
  /** @type {string[]} */
  const out = [];
  while (h.length) {
    const u = pop();
    out.push(names[u]);
    for (const v of g[u]) if (--deg[v] === 0) push(v);
  }
  std.out.puts(out.length < n ? "INVALID CYCLE\n" : `ORDER ${out.join(" ")}\n`);
}

File: problems/019-reproducible-build-order/solutions/go/main.go
package main

import (
	"bufio"
	"container/heap"
	"fmt"
	"os"
	"sort"
)

type H struct {
	a []int
	n []string
}

func (h H) Len() int           { return len(h.a) }
func (h H) Less(i, j int) bool { return h.n[h.a[i]] < h.n[h.a[j]] }
func (h H) Swap(i, j int)      { h.a[i], h.a[j] = h.a[j], h.a[i] }
func (h *H) Push(x any)        { h.a = append(h.a, x.(int)) }
func (h *H) Pop() any          { x := h.a[len(h.a)-1]; h.a = h.a[:len(h.a)-1]; return x }
func main() {
	in := bufio.NewReader(os.Stdin)
	var n, m int
	fmt.Fscan(in, &n, &m)
	names := make([]string, n)
	for i := range names {
		fmt.Fscan(in, &names[i])
	}
	nameOrder := make([]int, n)
	for i := range nameOrder {
		nameOrder[i] = i
	}
	sort.Slice(nameOrder, func(i, j int) bool { return names[nameOrder[i]] < names[nameOrder[j]] })
	findID := func(name string) int {
		i := sort.Search(n, func(i int) bool { return names[nameOrder[i]] >= name })
		if i < n && names[nameOrder[i]] == name {
			return nameOrder[i]
		}
		return -1
	}
	g := make([][]int, n)
	deg := make([]int, n)
	bad := 0
	for i := 1; i <= m; i++ {
		var a, b string
		fmt.Fscan(in, &a, &b)
		x, y := findID(a), findID(b)
		if x < 0 || y < 0 {
			if bad == 0 {
				bad = i
			}
		} else {
			g[y] = append(g[y], x)
			deg[x]++
		}
	}
	if bad > 0 {
		fmt.Println("INVALID DANGLING", bad)
		return
	}
	h := &H{n: names}
	heap.Init(h)
	for i := range names {
		if deg[i] == 0 {
			heap.Push(h, i)
		}
	}
	out := []string{}
	for h.Len() > 0 {
		u := heap.Pop(h).(int)
		out = append(out, names[u])
		for _, v := range g[u] {
			deg[v]--
			if deg[v] == 0 {
				heap.Push(h, v)
			}
		}
	}
	w := bufio.NewWriter(os.Stdout)
	defer w.Flush()
	if len(out) < n {
		fmt.Fprintln(w, "INVALID CYCLE")
	} else {
		fmt.Fprint(w, "ORDER")
		for _, x := range out {
			fmt.Fprint(w, " ", x)
		}
		fmt.Fprintln(w)
	}
}
